core-bundle/src/Controller/ContentElement/DownloadsController.php

```diff
@@ -0,0 +1,245 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Controller\ContentElement;
+
+use Contao\Config;
+use Contao\ContentModel;
+use Contao\CoreBundle\File\MetadataBag;
+use Contao\CoreBundle\Filesystem\FileDownloadHelper;
+use Contao\CoreBundle\Filesystem\FilesystemItem;
+use Contao\CoreBundle\Filesystem\FilesystemItemIterator;
+use Contao\CoreBundle\Filesystem\FilesystemUtil;
+use Contao\CoreBundle\Filesystem\SortMode;
+use Contao\CoreBundle\Filesystem\VirtualFilesystem;
+use Contao\CoreBundle\Image\Preview\MissingPreviewProviderException;
+use Contao\CoreBundle\Image\Preview\PreviewFactory;
+use Contao\CoreBundle\Image\Preview\UnableToGeneratePreviewException;
+use Contao\CoreBundle\Image\Studio\Figure;
+use Contao\CoreBundle\Image\Studio\Studio;
+use Contao\CoreBundle\ServiceAnnotation\ContentElement;
+use Contao\CoreBundle\Twig\FragmentTemplate;
+use Contao\FrontendUser;
+use Contao\LayoutModel;
+use Contao\StringUtil;
+use Symfony\Component\Filesystem\Path;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Security\Core\Security;
+
+/**
+ * @ContentElement("download", category="files")
+ * @ContentElement("downloads", category="files")
+ */
+class DownloadsController extends AbstractContentElementController
+{
+    private const DOWNLOAD_ROUTE_NAME = 'contao_download';
+
+    public function __construct(
+        private readonly Security $security,
+        private readonly VirtualFilesystem $filesStorage,
+        private readonly FileDownloadHelper $fileDownloadHelper,
+        private readonly PreviewFactory $previewFactory,
+        private readonly Studio $studio,
+        private readonly string $projectDir,
+        private readonly array|null $allowedFileExtensions = null,
+    ) {
+    }
+
+    /**
+     * This endpoint allows to stream files via PHP. This is for instance
+     * needed if a file has no public URI.
+     */
+    #[Route('/_download', name: self::DOWNLOAD_ROUTE_NAME)]
+    public function downloadAction(Request $request): Response
+    {
+        $this->initializeContaoFramework();
+
+        return $this->fileDownloadHelper->handle(
+            $request,
+            $this->filesStorage,
+            function (FilesystemItem $item, array $context): Response|null {
+                // TODO: What's needed here in terms of permissions?
```

Comment:
This needs to be addressed before we can merge.

Reply:
/cc @Toflar 🙏